policy-controller/grpc/src/lib.rs

```diff
@@ -134,8 +140,122 @@ where
     }
 }
 
+impl<T> OutboundPolicyServer<T>
+where
+    T: DiscoverOutboundPolicy<(String, String, NonZeroU16)> + Send + Sync + 'static,
+{
+    pub fn new(discover: T, drain: drain::Watch) -> Self {
+        Self {
+            index: discover,
+            drain,
+        }
+    }
+
+    pub fn svc(self) -> OutboundPoliciesServer<Self> {
+        OutboundPoliciesServer::new(self)
+    }
+
+    fn service_lookup(
+        &self,
+        target: TcpAddress,
+    ) -> Result<(String, String, NonZeroU16), tonic::Status> {
+        let port = target
+            .port
+            .try_into()
+            .map_err(|_| tonic::Status::invalid_argument("port outside valid range"))?;
+        let port = NonZeroU16::new(port)
+            .ok_or_else(|| tonic::Status::invalid_argument("port cannot be zero"))?;
+
+        let addr = target
+            .ip
+            .ok_or_else(|| tonic::Status::invalid_argument("traffic target must have an IP"))?
+            .try_into()
+            .map_err(|error| {
+                tonic::Status::invalid_argument(format!("failed to parse target addr: {}", error))
+            })?;
+
+        self.index
+            .service_lookup(addr, port)
+            .ok_or_else(|| tonic::Status::not_found("No such service"))
+    }
+}
+
+#[async_trait::async_trait]
+impl<T> OutboundPolicies for OutboundPolicyServer<T>
+where
+    T: DiscoverOutboundPolicy<(String, String, NonZeroU16)> + Send + Sync + 'static,
+{
+    async fn get(
+        &self,
+        req: tonic::Request<outbound::TrafficSpec>,
+    ) -> Result<tonic::Response<outbound::OutboundPolicy>, tonic::Status> {
+        let traffic_spec = req.into_inner();
+
+        let target = traffic_spec
+            .target
+            .ok_or_else(|| tonic::Status::invalid_argument("target is required"))?;
+        let target = match target {
+            outbound::traffic_spec::Target::Addr(target) => target,
+            outbound::traffic_spec::Target::Authority(_) => {
+                return Err(tonic::Status::unimplemented(
+                    "getting policy by authority is not supported",
+                ))
+            }
+        };
```

Comment:
nit, not a blocker: seems like this could be part of `service_lookup` or something, since it's the same in both `get` and `watch`?